File: Python3/0897. Increasing Order Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # def increasingBST(self, root: TreeNode) -> TreeNode:
    #     def vistAllLeft(root):
    #         res = []
    #         while root:
    #             res.append(root)
    #             root = root.left
            
    #         return res
        
    #     dummy = TreeNode(-1)
    #     temp = dummy
    #     stack = vistAllLeft(root)
        
    #     while stack:
    #         node = stack.pop()
    #         temp.right = TreeNode(node.val)
    #         temp = temp.right

    #         node = node.right
    #         stack += vistAllLeft(node)
        
    #     return dummy.right
    
    
    # def increasingBST(self, root: TreeNode) -> TreeNode:
    #     if not root:
    #         return

    #     temp = dummy = TreeNode(-1)

    #     temp.right = self.increasingBST(root.left)

    #     while temp.right:
    #         temp = temp.right
        
    #     temp.right = TreeNode(root.val)
        
    #     temp.right.right = self.increasingBST(root.right)

    #     return dummy.right
    

    def increasingBST(self, root: TreeNode) -> TreeNode:
        if not root:
            return
        
        temp = ans = TreeNode(-1)
        temp.right = self.increasingBST(root.left)

        while temp.right:
            temp = temp.right
        
        root.left = None
        temp.right = root

        root.right = self.increasingBST(root.right)

        return ans.right

        

File: C++/0897. Increasing Order Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        TreeNode* res = new TreeNode(-1);
        TreeNode* temp = res;
        stack<TreeNode*> arr;
        TreeNode* node = root;

        while(node != nullptr) {
            arr.push(node);
            node = node->left;
        }

        while(!arr.empty()) {
            node = arr.top();
            arr.pop();
            temp->right = new TreeNode(node->val);
            temp = temp->right;
            if(node->right != nullptr) {
                node = node->right;
                while(node != nullptr) {
                    arr.push(node);
                    node = node->left;
                }
            }
        }
        return res->right;

    }
};
